elastic/kibana
# [HTTP] Add support for configuring a CDN (part I)

## Summary

This PR prepares the Kibana core code for configuring and using a CDN to serve the vast majority of assets.

## Goals

(1) Configure a CDN URL: `server.cdn.url`
(2) Reach a new domain: updating our CSP to allow this
(3) Serve JS bundles, CSS, fonts via CDN

## How to test

1. Make a release/distribution build of Kibana (`node ./scripts/build.js --release --skip-os-packages`), will output to `./build/kibana`.
2. Edit `/etc/hosts`
    1. I added `127.0.0.1 my.cdn.test`
3. Edit `config/kibana.dev.yml` adding `server.cdn.url: "http://my.cdn.test:1772"`
4. `cd` into `build/kibana` once the release build is done
5. Download and run this (hacky) script there. It should (🤞🏻 ) create a folder structure mirroring our server paths: https://gist.github.com/jloleysens/86cae7ea3d7512e6a6e92cebbf1566f4
6. Start a file server in `build/kibana`. For example: `npx http-server -p 1772 --cors --gzip --brotli` -- this will be the "CDN" server
7. Start Kibana `yarn start` and ES `yarn es snapshot`
8. Open Kibana, checking the Network tab of your browser for anything that fails to load: `404`s, CORs or CSP is almost definitely a bug in this code.

## Future work
* Include the translations file to the CDN (~should not be much effort~)
* Update the build process to create a folder structure that mirrors what we serve from Kibana today, this will replace step 5 in the testing process
  * This includes deploying new assets to the CDN as part of our promotion pipeline


## Screenshots

<img width="1391" alt="Screenshot 2023-10-19 at 17 38 56" src="https://github.com/elastic/kibana/assets/8155004/1e0272b7-d117-4ef2-a834-a5a163d6e4cd">

### Checklist

Delete any items that are not applicable to this PR.



### Risk Matrix

Delete this section if it is not applicable to this PR.

Before closing this PR, invite QA, stakeholders, and other developers to identify risks that should be tested prior to the change/feature release.

When forming the risk matrix, consider some of the following examples and how they may potentially impact the change:

| Risk                      | Probability | Severity | Mitigation/Notes        |
|---------------------------|-------------|----------|-------------------------|
| Changing the existing behaviour | med | high | based on the existing test coverage I expect us to not change the current non-CDN asset serving  |
| Use of undocumented feature | Very low | Low | While this could brick Kibana, it would be a simple configuration change  |


### For maintainers
